src/domain_bridge/wait_for_qos_handler.hpp

```diff
@@ -0,0 +1,206 @@
+   * \param callback: User callback that is triggered when QoS settings can be determined
+   *   for the provided topic name.
+   *   If QoS settings are already available, then the callback is called immediately.
+   */
+  void register_on_publisher_qos_ready_callback(
+    std::string topic,
+    rclcpp::Node::SharedPtr node,
+    std::function<void(const QosMatchInfo)> callback)
+  {
+    // If the QoS is already avaiable, trigger the callback immediately
+    auto opt_qos = this->get_topic_qos(topic, node);
+    if (opt_qos) {
+      const QosMatchInfo & qos = opt_qos.value();
+      callback(qos);
+      return;
+    }
+
+    // Create a thread that waits for a publisher to become available
+    auto invoke_callback_when_qos_ready = [this, topic, node, callback]()
+      {
+        auto event = node->get_graph_event();
+        while (!this->shutting_down_.load()) {
+          // Wait for graph event
+          // Note, in case new publishers don't trigger a graph event we add a
+          // timeout so that we can still poll periodically for new publishers
+          node->wait_for_graph_change(event, std::chrono::seconds(1));
+          event->check_and_clear();
+          // Check if QoS is ready
+          auto opt_qos = this->get_topic_qos(topic, node);
+          if (opt_qos) {
+            const QosMatchInfo & qos = opt_qos.value();
+            callback(qos);
+            return;
+          }
+        }
+      };
+    auto waiting_thread = std::make_shared<std::thread>(invoke_callback_when_qos_ready);
+    waiting_threads_.push_back(waiting_thread);
+  }
+
+private:
+  /// Get QoS settings that best match all available publishers.
+  /**
+   * Queries QoS of existing publishers and returns QoS that is compatibile
+   * with the majority of publishers.
+   * If possible, the returned QoS will exactly match those of the publishers.
+   *
+   * If the existing publishers do not have matching QoS settings, then a warning is set in the
+   * return struct.
+   *
+   * If there are no publishers, then no QoS is returned (i.e. the optional object is not set).
+   */
+  std::optional<QosMatchInfo> get_topic_qos(std::string topic, rclcpp::Node::SharedPtr node)
+  {
+    // TODO(jacobperron): replace this with common implementation when it is available.
+    //       See: https://github.com/ros2/rosbag2/issues/601
+
+    // This implementation is inspired by rosbag2_transport:
+    // https://github.com/ros2/rosbag2/blob/master/rosbag2_transport/src/rosbag2_transport/qos.cpp
+
+    // Query QoS info for publishers
+    std::vector<rclcpp::TopicEndpointInfo> endpoint_info_vec =
+      node->get_publishers_info_by_topic(topic);
+    std::size_t num_endpoints = endpoint_info_vec.size();
+
+    // If there are no publishers, return default QoS
+    if (num_endpoints < 1u) {
+      return {};
+    }
+
+    // Initialize QoS arbitrarily
+    QosMatchInfo result_qos(endpoint_info_vec[0].qos_profile());
```

Comment:
```suggestion
    QosMatchInfo result_qos;
    result_qos.qos.reliability(endpoint_info_vec[0].qos_profile().reliability());
    result_qos.qos.durability(endpoint_info_vec[0].qos_profile().durability());
```

I think copying the liveliness and deadline of the read profile isn't a good idea.
Each of the publishers might have a different one, but the good thing is that if your subscription has the default liveliness and default deadline it will match everything.

Reply:
Hmmm, yeah good point. We could also take the largest of all values for deadline and lifespan (note the QoS here is also applied to the publisher). Liveliness is trickier; I'm not sure how to handle the "manual by topic" since we are republishing to the other side. Maybe it's best to always use "automatic" for the bridge publisher. What do you think?

Reply:
>  I'm not sure how to handle the "manual by topic" since we are republishing to the other side

A subscription with "AUTOMATIC" liveliness will match everything, and the publisher should also be automatic if not you will need to have code doing some manual `assert_liveliness` calls.

> We could also take the largest of all values for deadline and lifespan

Sounds fine to me, that's in general the default but manuall setting the durations to "INFINITY" is better.

Reply:
> Sounds fine to me, that's in general the default but manuall setting the durations to "INFINITY" is better.

I'd like to try and mimic the QoS as best we can, so I'll avoid setting the durations to infinity by default.

I think we have to make an exception for liveliness, and explicitly use AUTOMATIC.

Reply:
See c62b3a1

- Always use automatic liveliness
- Use max of all deadline
- Use max of all lifespan

Updated the design doc and added tests.

Reply:
> Use max of all deadline
> Use max of all lifespan

IMO, those two are better manually chosen, "replicating" them doesn't seem to completely make sense.

Reply:
They seem to make sense to me. If we have a publisher in domain A with a deadline, then using the same deadline for the bridge publisher on domain B will unsure subscriptions in domain B will experience similar behavior as subscriptions in domain A. Ditto for lifespan. What in particular doesn't make sense about this logic?

I will definitely expose a configuration point for users to override QoS values (e.g. via the YAML config). Expect a follow-up PR after this one.

Reply:
> They seem to make sense to me. If we have a publisher in domain A with a deadline, then using the same deadline for the bridge publisher on domain B will ensure subscriptions in domain B will experience similar behavior as subscriptions in domain A. 

The thing is that the bridge will introduce some delay, so you tipically won't be able to met the same deadline as the original publisher.
If deadlines are important, I would rather think what's a reasonable deadline for the new domain than letting the bridge infer them.

> Ditto for lifespan. What in particular doesn't make sense about this logic?

Lifespan is how much the message lives in the queue, and it's important how that's is combined with the history size/kind.
Because the last two cannot be instronspected, I don't think that inferring the lifespan will always lead to good results.

---

anyways, keeping this logic sounds fine to me.

Reply:
> The thing is that the bridge will introduce some delay, so you tipically won't be able to met the same deadline as the original publisher.

Since deadline is a duration between messages, any delay introduced by the bridge should (hopefully) be systematic. Even though the total time from the original publisher to the endpoint in the another domain may take longer, the time between consecutive messages republished by the bridged should ideally remain the same as they were before they entered the bridge. E.g. if the original publisher is publishing at 10Hz, I would expect the bridge to also publish at 10Hz into the other domain.

> Lifespan is how much the message lives in the queue, and it's important how that's is combined with the history size/kind.

This is a good point. I can add a note about this potential pitfall in the design doc.

Reply:
I've reconsidered and changed the default values to be max integers instead of matching available publishers in https://github.com/ros2/domain_bridge/pull/13